src/sql/src/query_model/mod.rs

```diff
@@ -398,6 +400,72 @@ impl QueryBox {
         }
     }
 
+    /// Returns the type of a given column projected by this box.
+    fn column_type(&self, model: &Model, column: usize) -> ColumnType {
+        let the_expr = &self.columns[column].expr;
+        let mut column_type = the_expr.typ(model);
+        match &self.box_type {
+            BoxType::Select(select) => {
+                if column_type.nullable {
+                    // Check whether the box contains predicates that reject
+                    // nulls for all the non null requeriments of the expression.
+                    let mut non_null_requirements = HashSet::new();
+                    the_expr.non_null_requirements(&mut non_null_requirements);
+                    let mut non_null_columns = HashSet::new();
+                    for predicate in select.predicates.iter() {
```

Comment:
I don't think we should be calling `BoxScalarExpr::non_null_requirements` on the predicate expressions in order to infer columns that will be rejected if null.

For example, for the `x IS NOT NULL` fragment in

```sql
SELECT coalesce(x, y) FROM T WHERE x IS NOT NULL
```
the corresponding `non_null_requirements` returns an empty set, whereas I would intuitively expect it to return `x`.